[Curl] Reduce unnecessary SharedBuffer creation by making CurlMultipartHandle::didReceiveMessage take std::span<uint8_t>

kshukuwa opened the pull request:
#### 7ab6aab9c1b7e6f09edf3d7ceed204070132a7ce
<pre>
[Curl] Reduce unnecessary SharedBuffer creation by making CurlMultipartHandle::didReceiveMessage take std::span&lt;uint8_t&gt;
<a href="https://bugs.webkit.org/show_bug.cgi?id=262472">https://bugs.webkit.org/show_bug.cgi?id=262472</a>

Reviewed by Fujii Hironori.

Currently, a SharedBuffer is generated in CurlRequest::didReceiveDataCallback
when receiving data from libcurl. However, if CurlRequest::didReceiveData
performs CURL_WRITEFUNC_PAUSE processing, this SharedBuffer will be
discarded without being used.

Also, CurlMultipartHandle::didReceiveMessage performs a SharedBuffer to
Vector conversion. This is a wasteful data conversion.

For this reason, avoid converting to SharedBuffer at the time of
CurlRequest::didReceiveDataCallback.

* Source/WebCore/platform/network/curl/CurlMultipartHandle.cpp:
(WebCore::CurlMultipartHandle::didReceiveMessage):
(WebCore::CurlMultipartHandle::processContent):
* Source/WebCore/platform/network/curl/CurlMultipartHandle.h:
* Source/WebCore/platform/network/curl/CurlMultipartHandleClient.h:
* Source/WebCore/platform/network/curl/CurlRequest.cpp:
(WebCore::CurlRequest::didReceiveData):
(WebCore::CurlRequest::didReceiveDataFromMultipart):
(WebCore::CurlRequest::didReceiveDataCallback):
* Source/WebCore/platform/network/curl/CurlRequest.h:
* Source/WebCore/platform/network/curl/CurlRequestClient.h:
* Source/WebKit/NetworkProcess/curl/NetworkDataTaskCurl.cpp:
(WebKit::NetworkDataTaskCurl::curlDidReceiveData):
* Source/WebKit/NetworkProcess/curl/NetworkDataTaskCurl.h:
* Tools/TestWebKitAPI/Tests/WebCore/curl/CurlMultipartHandleTests.cpp:
(TestWebKitAPI::Curl::toSpan):
(TestWebKitAPI::Curl::TEST):

Canonical link: <a href="https://commits.webkit.org/268801@main">https://commits.webkit.org/268801@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/af43df723288cbe5b2556cfe02090734a504fa98

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/20647 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/21065 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/21720 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/22541 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/19270 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/20884 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/24319 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/21246 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/20604 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/20869 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20695 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17949 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/23397 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17861 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/18755 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/25022 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18940 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/18932 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/22966 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19520 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/16583 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18740 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/23077 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2557 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19338 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

fujii commented on Source/WebCore/platform/network/curl/CurlMultipartHandle.cpp:
```diff
@@ -79,12 +79,12 @@ CurlMultipartHandle::CurlMultipartHandle(CurlMultipartHandleClient& client, CStr
 {
 }
 
-void CurlMultipartHandle::didReceiveMessage(const SharedBuffer& buffer)
```
Recently, WebKit is using `std::span<uint8_t>` for this purpose.

kshukuwa replied:
Fixed

fujii reviewed (changes requested):
> [Curl] Reduce unnecessary data conversions

This doesn't seems to be informative enough to me. How about "[Curl] Reduce unnecessary SharedBuffer creation by making CurlMultipartHandle::didReceiveMessage take std::span<uint8_t>"

fujii commented on Source/WebCore/platform/network/curl/CurlMultipartHandle.cpp:
```diff
@@ -79,12 +79,12 @@ CurlMultipartHandle::CurlMultipartHandle(CurlMultipartHandleClient& client, CStr
         return; // The handler is closed down so ignore everything.
 
-    m_buffer.append(buffer.data(), buffer.size());
+    m_buffer.append(receivedData.data(), receivedData.size());
```
m_buffer.append(receivedData);

kshukuwa replied:
Fixed

fujii commented on Source/WebCore/platform/network/curl/CurlMultipartHandle.cpp:
```diff
@@ -79,12 +79,12 @@ CurlMultipartHandle::CurlMultipartHandle(CurlMultipartHandleClient& client, CStr
 }
 
-void CurlMultipartHandle::didReceiveMessage(const SharedBuffer& buffer)
+void CurlMultipartHandle::didReceiveMessage(std::span<const uint8_t>&& receivedData)
```
Grepping `std::span` through WebKit, you should remove `&&` here.

kshukuwa replied:
Fixed

fujii commented on Tools/TestWebKitAPI/Tests/WebCore/curl/CurlMultipartHandleTests.cpp:
```diff
@@ -782,7 +782,7 @@ TEST(CurlMultipartHandleTests, MaxHeaderSize)
     auto handle = CurlMultipartHandle::createIfNeeded(client, curlResponse);
 
-    handle->didReceiveMessage(SharedBuffer::create(data.data(), data.size()));
+    handle->didReceiveMessage({ data.data(), data.size() });
```
handle->didReceiveMessage(data.span());

kshukuwa replied:
Fixed

fujii commented on Tools/TestWebKitAPI/Tests/WebCore/curl/CurlMultipartHandleTests.cpp:
```diff
@@ -139,7 +139,7 @@ TEST(CurlMultipartHandleTests, SimpleMessage)
     auto handle = CurlMultipartHandle::createIfNeeded(client, curlResponse);
 
-    handle->didReceiveMessage(SharedBuffer::create(data, strlen(data)));
+    handle->didReceiveMessage({ reinterpret_cast<const uint8_t*>(data.characters()), data.length() });
```
How about defining an function converting a `ASCIILiteral` to `std::span<const uint8_t>` in this file?
std::span<const uint8_t> toSpan(const ASCIILiteral& x);

kshukuwa replied:
Fixed

kshukuwa committed:
[Curl] Reduce unnecessary SharedBuffer creation by making CurlMultipartHandle::didReceiveMessage take std::span<uint8_t>
https://bugs.webkit.org/show_bug.cgi?id=262472

Reviewed by Fujii Hironori.

Currently, a SharedBuffer is generated in CurlRequest::didReceiveDataCallback
when receiving data from libcurl. However, if CurlRequest::didReceiveData
performs CURL_WRITEFUNC_PAUSE processing, this SharedBuffer will be
discarded without being used.

Also, CurlMultipartHandle::didReceiveMessage performs a SharedBuffer to
Vector conversion. This is a wasteful data conversion.

For this reason, avoid converting to SharedBuffer at the time of
CurlRequest::didReceiveDataCallback.

* Source/WebCore/platform/network/curl/CurlMultipartHandle.cpp:
(WebCore::CurlMultipartHandle::didReceiveMessage):
(WebCore::CurlMultipartHandle::processContent):
* Source/WebCore/platform/network/curl/CurlMultipartHandle.h:
* Source/WebCore/platform/network/curl/CurlMultipartHandleClient.h:
* Source/WebCore/platform/network/curl/CurlRequest.cpp:
(WebCore::CurlRequest::didReceiveData):
(WebCore::CurlRequest::didReceiveDataFromMultipart):
(WebCore::CurlRequest::didReceiveDataCallback):
* Source/WebCore/platform/network/curl/CurlRequest.h:
* Source/WebCore/platform/network/curl/CurlRequestClient.h:
* Source/WebKit/NetworkProcess/curl/NetworkDataTaskCurl.cpp:
(WebKit::NetworkDataTaskCurl::curlDidReceiveData):
* Source/WebKit/NetworkProcess/curl/NetworkDataTaskCurl.h:
* Tools/TestWebKitAPI/Tests/WebCore/curl/CurlMultipartHandleTests.cpp:
(TestWebKitAPI::Curl::toSpan):
(TestWebKitAPI::Curl::TEST):

Canonical link: https://commits.webkit.org/268801@main
